README.rst
Commandify
==========

Allows simple creation of Python command line utilities through decorating functions. Builds on the ``argparse`` module in the standard library. Optionally uses the ``argcomplete`` module to provide command line tab-completion for bash/zsh.

Installation
============

::

    pip install commandify

Running example and usage
=========================

Once commandify has been installed, it can be run with:

::

    commandify_examples --help


``commandify_examples`` code:

.. code:: python
    
    #!/usr/bin/env python
    # Only required if you want to use argcomplete:
    # PYTHON_ARGCOMPLETE_OK
    '''Examples of a simple set of functions that use commandify
        usage::
    
            commandify_examples --help
            commandify_examples <command> --help
            commandify_examples --main-arg=22 <command>
            commandify_examples -m=22 <command>
            commandify_examples cmd_no_args
    
    '''
    from commandify import commandify, command, main_command
    
    
    @main_command(main_arg={'flag': '-m'})
    def main(main_arg=345):
        '''Example of how to use commandify
        Main command, called before any other command
        Can be used to set any global variables before any other commands executed.
        usage::
    
            commandify_examples --help
            commandify_examples --main-arg=22 cmd_no_args
            commandify_examples -m=22 cmd_no_args
            commandify_examples cmd_no_args
    
        '''
        print(type(main_arg))
        print("Main command: {0}".format(main_arg))
    
    
    @command
    def cmd_no_args():
        '''Simplest command
        usage::
    
            commandify_examples cmd_no_args
    
        '''
        print('cmd_no_args called')
    
    
    @command
    def cmd1(name):
        '''Simple command with argument
        usage::
    
            commandify_examples cmd1 --name=a1
            commandify_examples cmd1 --name a3
            commandify_examples cmd1 --name 'a4'
    
        '''
        print('cmd1 running')
        print(name)
    
    
    @command(username={'flag': '-n'})
    def cmd2(username, userid):
        '''Command with 2 args and a short flag set for name
        usage::
    
            commandify_examples cmd2 --username=steve --userid=55
            commandify_examples cmd2 -n=sarah --userid=56
    
        '''
        print('{0}: {1}'.format(userid, username))
    
    
    @command(some_arg={'default': 377})
    def cmd3(some_arg, arg_with_default='arg_default'):
        '''Command with defaults set in two ways
        first through decorator and second through function arguments.
        usage::
    
            commandify_examples cmd3
            commandify_examples cmd3 --arg-with-default=default_overridden
    
        '''
        # Type of some_arg will be int:
        print(type(some_arg))
        print(some_arg, arg_with_default)
    
    
    @command
    def cmd4(some_arg=False):
        '''Command with False bool default
        usage::
    
            commandify_examples cmd4
            commandify_examples cmd4 --some-arg
    
        '''
        # Type of some_arg will be bool:
        print(type(some_arg))
        # If it is called specifying --some-arg, it will be True.
        print(some_arg)
    
    
    @command
    def cmd5(some_arg=True):
        '''Command with True bool default
        Command line argument gets turned into negative to handle this.
        usage::
    
            commandify_examples cmd5
            commandify_examples cmd5 --not-some-arg
    
        '''
        # Type of some_arg will be bool:
        print(type(some_arg))
        # If it is called without specifying --not-some-arg, it will be True.
        print(some_arg)
    
    
    @command
    def cmd6():
        print('cmd6 called')
    
    
    if __name__ == '__main__':
        try:
            import argcomplete
            commandify(suppress_warnings=['default_true'], use_argcomplete=True)
        except ImportError:
            commandify(suppress_warnings=['default_true'])


docs/changelog.rst
Change Log
==========

Version 0.0.4.5 (Alpha) - May 17, 2015
--------------------------------------

* Add support for ``argcomplete``

Version 0.0.4.3 (Alpha) - May 15, 2015
--------------------------------------

* Add in unit tests for main command only
* Change handling of ``parse_args`` in main parser so that the NS returned 
  has had e.g. ``not_some_arg=True`` replaced by ``some_arg=False``

Version 0.0.4.2 (Alpha) - May 15, 2015
--------------------------------------

* Fix serious bug that meant that all local vars in func were counted as arguments
* Auto setting of type=<type> for default args based on default's type
* Handle some_arg=[True|False] in natural way

Version 0.0.4.1 (Alpha) - May 14, 2015
--------------------------------------

* Change ``VERSION.txt`` to ``commandify/version.py``
* Allow for providing arguments for all functions
* Update setup.py to be able to run tests

Version 0.0.3.1 (Alpha) - May 13, 2015
--------------------------------------

* Switch to subclassing ``argparse.ArgumentParser`` to allow more flexibility.
* Allow for providing arguments for all functions

Version 0.0.2.2 (Alpha) - May 13, 2015
--------------------------------------

* Initial Python3 support
* Added example command (in docstrings) test
* Fixed some bad example commands
* Added a release procedure
* Allowed ``_commands`` and ``_main_command`` to be accessed though module

Version 0.0.1 (Alpha) - May 12, 2015
------------------------------------

* Basic functionality in place
    * Decorators for functions converted to command line args
    * Adding options to decorators
* Python project started
* Uploaded to PyPI



tests/unit/unit_tests.py
# Borrows heavily from argparse tests.
# Doesn't use meta class to setup tests.
import sys
import unittest
from collections import OrderedDict

import commandify as cmdify
print(cmdify._main_commands)

try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO


class StdIOBuffer(StringIO):
    pass


class NS(object):
    def __init__(self, **kwargs):
        self.__dict__.update(kwargs)

    def __repr__(self):
        sorted_items = sorted(self.__dict__.items())
        kwarg_str = ', '.join(['%s=%r' % tup for tup in sorted_items])
        return '%s(%s)' % (type(self).__name__, kwarg_str)

    def __eq__(self, other):
        return vars(self) == vars(other)

    def __ne__(self, other):
        return not (self == other)


class ArgumentParserError(Exception):
    def __init__(self, message, stdout=None, stderr=None, error_code=None):
        Exception.__init__(self, message, stdout, stderr)
        self.message = message
        self.stdout = stdout
        self.stderr = stderr
        self.error_code = error_code


def turn_system_exit_into_error(func, *args, **kwargs):
    ### TAKEN FROM argparse ###
    # if called recursively just calls func
    # else redirects stdout, stderr and calls func, catching
    # and SystemExit exceptions. Always reconnects stdout/stderr
    # Don't fully understand - particularly commanted out part.

    # if this is being called recursively and stderr or stdout is already being
    # redirected, simply call the function and let the enclosing function
    # catch the exception
    if isinstance(sys.stderr, StdIOBuffer) or isinstance(sys.stdout, StdIOBuffer):
        return func(*args, **kwargs)

    # if this is not being called recursively, redirect stderr and
    # use it as the ArgumentParserError message
    old_stdout = sys.stdout
    old_stderr = sys.stderr
    sys.stdout = StdIOBuffer()
    sys.stderr = StdIOBuffer()

    try:
        try:
            result = func(*args, **kwargs)
            # Not sure what this does?
            # for key in list(vars(result)):
            #     if getattr(result, key) is sys.stdout:
            #         setattr(result, key, old_stdout)
            #     if getattr(result, key) is sys.stderr:
            #         setattr(result, key, old_stderr)
            return result
        except SystemExit:
            code = sys.exc_info()[1].code
            stdout = sys.stdout.getvalue()
            stderr = sys.stderr.getvalue()
            raise ArgumentParserError("SystemExit", stdout, stderr, code)
    finally:
        sys.stdout = old_stdout
        sys.stderr = old_stderr


class ErrorRaisingArgumentParser(cmdify.CommandifyArgumentParser):
    '''Allows testing of argument parser errors.
    
    Instead of raising SystemExit will raise ArgumentParserError.
    Performs some tricks to redirect stdout/stderr too.'''
    def parse_args(self, *args, **kwargs):
        parse_args = super(ErrorRaisingArgumentParser, self).parse_args
        return turn_system_exit_into_error(parse_args, *args, **kwargs)

    def exit(self, *args, **kwargs):
        exit = super(ErrorRaisingArgumentParser, self).exit
        return turn_system_exit_into_error(exit, *args, **kwargs)

    def error(self, *args, **kwargs):
        error = super(ErrorRaisingArgumentParser, self).error
        return turn_system_exit_into_error(error, *args, **kwargs)


class BaseUnitTest(unittest.TestCase):
    def setUp(self):
        self.parser = ErrorRaisingArgumentParser()
        cmdify._main_commands.clear()
        cmdify._commands.clear()


    def _run_dispatch_tests(self, failures, successes):
        self.parser.setup_arguments()
        self._test_failures(failures)
        self._test_dispatch_successes(successes)

    def _test_failures(self, failures):
        for args in failures:
            print(args)
            if isinstance(args, tuple):
                test_ns = True
                args, expected_ns = args[0], args[1]
            else:
                test_ns = False

            if isinstance(args, str):
                args = args.split()

            if test_ns:
                result_ns = self.parser.parse_args(args)
                print(result_ns, expected_ns)
                assert result_ns != expected_ns
            else:
                raises = self.assertRaises
                raises(ArgumentParserError, self.parser.parse_args, args)

    def _test_dispatch_successes(self, successes):
        for args, expected_ns, expected_ret in successes:
            print(args)
            if isinstance(args, str):
                args = args.split()

            result_ns = self.parser.parse_args(args)
            print(result_ns, expected_ns)
            assert result_ns == expected_ns
            result_ret = self.parser.dispatch_commands()
            print(result_ret, expected_ret)
            assert result_ret == expected_ret


count = 0

class TestMainDispatch(BaseUnitTest):
    def test_1_blank(self):
        start_count = count

        @cmdify.main_command
        def m():
            global count
            count += 1
            return None

        failures = ['f', 'arg', '-f', '--f']

        successes = [
            ('', NS(), (None, None)),
            (' ', NS(), (None, None)),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count

    def test_2_single_arg(self):
        start_count = count

        @cmdify.main_command
        def m(arg):
            global count
            count += 1
            assert arg == '5'
            return arg

        failures = ['f', 'arg', '-f', '--f', '--arg5']

        successes = [
            ('--arg 5', NS(arg='5'), ('5', None)),
            ('--arg=5', NS(arg='5'), ('5', None)),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count

    def test_3_single_arg_with_default(self):
        start_count = count
        @cmdify.main_command
        def m(arg=5):
            global count
            count += 1
            return arg

        failures = ['f', 'arg', '-f', '--f', '--arg5', '--arg five', '--arg',
            '--arg=5.4',
            ('', NS(arg=6)),
            ('', NS(arg='5')),
            ('--arg=5', NS(arg='5')),
        ]

        successes = [
            ('', NS(arg=5), (5, None)),
            ('--arg=7', NS(arg=7), (7, None)),
            ('--arg 7', NS(arg=7), (7, None)),
            ('--arg=-7', NS(arg=-7), (-7, None)),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count

    def test_4_single_arg_with_flag(self):
        start_count = count
        @cmdify.main_command(arg={'flag': '-a'})
        def m(arg):
            global count
            count += 1
            return arg

        failures = [' ', 'f', 'arg', '-f', '--f', '--arg5', '--arg', '-a',
            ('-a 7', NS(arg=7)),
            ('-a-999', NS(arg=-999)),
        ]

        successes = [
            ('--arg=5', NS(arg='5'), ('5', None)),
            ('--arg 7', NS(arg='7'), ('7', None)),
            ('--arg=-5', NS(arg='-5'), ('-5', None)),
            ('-a=5', NS(arg='5'), ('5', None)),
            ('-a7', NS(arg='7'), ('7', None)),
            ('-a -8', NS(arg='-8'), ('-8', None)),
            ('-a 999', NS(arg='999'), ('999', None)),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count

    def test_5_single_arg_with_default_and_flag(self):
        start_count = count
        @cmdify.main_command(arg={'flag': '-a'})
        def m(arg=1234):
            global count
            count += 1
            return arg

        failures = ['f', 'arg', '-f', '--f', '--arg5', '--arg', '-a',
            ('-a 7', NS(arg='7')),
            ('-a9', NS(arg='9')),
        ]

        successes = [
            ('', NS(arg=1234), (1234, None)),
            ('--arg=5', NS(arg=5), (5, None)),
            ('--arg 7', NS(arg=7), (7, None)),
            ('--arg=-5', NS(arg=-5), (-5, None)),
            ('-a=5', NS(arg=5), (5, None)),
            ('-a7', NS(arg=7), (7, None)),
            ('-a -8', NS(arg=-8), (-8, None)),
            ('-a 999', NS(arg=999), (999, None)),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count

    def test_6_single_arg_with_bool_false(self):
        start_count = count
        @cmdify.main_command
        def m(arg=False):
            global count
            count += 1
            return arg

        failures = ['f', 'arg', '-f', '--f', '--arg5', '-a',
            ('--arg', NS(arg=False)),
        ]

        successes = [
            ('', NS(arg=False), (False, None)),
            ('--arg', NS(arg=True), (True, None)),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count

    def test_7_single_arg_with_bool_true(self):
        start_count = count
        @cmdify.main_command
        def m(arg=True):
            global count
            count += 1
            return arg

        failures = ['f', 'arg', '-f', '--f', '--arg5', '--arg', '-a',
            ('--not-arg', NS(arg=True)),
        ]

        successes = [
            ('', NS(arg=True), (True, None)),
            ('--not-arg', NS(arg=False), (False, None)),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count

    def test_8_access_args(self):
        start_count = count
        @cmdify.main_command
        def m(args, some_arg):
            global count
            count += 1
            return args.some_arg

        failures = ['f', 'args', '--args', '--arg', '-a']

        successes = [
            ('--some-arg=sarah', NS(some_arg='sarah'), ('sarah', None)),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count

class TestCommandDispatch(BaseUnitTest):
    def test_1_blank(self):
        start_count = count

        @cmdify.main_command
        def m():
            return None

        @cmdify.command
        def c():
            global count
            count += 1
            return None


        failures = ['f', 'arg', '-f', '--f']

        successes = [
            ('c', NS(command='c'), (None, None)),
            ('  c  ', NS(command='c'), (None, None)),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count

    def test_2_required_arg(self):
        start_count = count

        @cmdify.main_command
        def m():
            return None

        @cmdify.command
        def c(some_arg):
            global count
            count += 1
            return some_arg

        failures = ['f', '--some-arg', '--some-arg c', 'c', 'c --some-arg',
            ('c --some-arg=4', NS(command='c', some_arg=4)),
        ]

        successes = [
            ('c --some-arg=4', NS(command='c', some_arg='4'), (None, '4')),
            ('c --some-arg=dave', NS(command='c', some_arg='dave'), (None, 'dave')),
            ('c --some-arg john', NS(command='c', some_arg='john'), (None, 'john')),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count


    def test_3_default_arg(self):
        start_count = count

        @cmdify.main_command
        def m():
            return None

        @cmdify.command
        def c(some_arg='jo'):
            global count
            count += 1
            return some_arg

        failures = ['f', '--jo', '--some-arg c', 'c --some-arg',
            ('c', NS(command='c', some_arg='bib')),
            ('c --some-arg=bob', NS(command='c', some_arg='bib')),
        ]

        successes = [
            ('c', NS(command='c', some_arg='jo'), (None, 'jo')),
            ('c --some-arg=4', NS(command='c', some_arg='4'), (None, '4')),
            ('c --some-arg=dave', NS(command='c', some_arg='dave'), (None, 'dave')),
            ('c --some-arg john', NS(command='c', some_arg='john'), (None, 'john')),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count

    def test_4_access_args(self):
        start_count = count

        @cmdify.main_command
        def m():
            return None

        @cmdify.command
        def c(args):
            global count
            count += 1
            return args.command

        failures = ['args', '--args', 'c --args', 'c --args=doug',
            ('c', NS(command='d')),
        ]

        successes = [
            ('c', NS(command='c'), (None, 'c')),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count

    def test_5_main_ret(self):
        start_count = count

        @cmdify.main_command
        def m(main_arg=9):
            return main_arg + 1

        @cmdify.command
        def c(args):
            global count
            count += 1
            return args.main_ret + 1

        failures = ['--main-arg=4', '--main-arg=fda c',
            ('c', NS(command='c')),
            ('c', NS(command='c', main_arg=10)),
        ]

        successes = [
            ('c', NS(command='c', main_arg=9), (10, 11)),
            ('--main-arg=22 c', NS(command='c', main_arg=22), (23, 24)),
        ]

        self._run_dispatch_tests(failures, successes)
        assert count == len(successes) + start_count


tests/pep8_tests/test_code_format.py
from glob import glob
import pep8


class TestCodeFormat(object):
    def __init__(self):
        self.pep8style = pep8.StyleGuide()

    def _test_conformance_in_files(self, filenames):
        assert len(filenames) != 0
        for filename in filenames:
            result = self.pep8style.check_files([filename])
            assert result.total_errors == 0, "Found code style errors (and warnings)."

    def test_1_commandify_pep8_conformance(self):
        """Test that commandify module conforms to PEP8."""
        filenames = glob('commandify/*.py')
        self._test_conformance_in_files(filenames)


tests/doc/test_documentation_example_commands.py
import inspect
import subprocess
import os

import commandify as cmdify


def parse_doc_for_commands(doc):
    if not doc:
        return []
    cmds = []
    lines = doc.split('\n')
    usage_found = False
    for line in lines:
        if line[:10] == '    usage:':
            usage_found = True
        if usage_found:
            if line[:27] == '        commandify_examples':
                cmds.append(parse_command(line))
    return cmds


def parse_command(line):
    cmd = line[8:]
    return './{0}'.format(cmd)


class TestDocumentationExampleCommands(object):
    def __init__(self):
        pass

    def test_1_test_all_example_commands(self):
        """Test that example commands execute and give return code 0"""
        cmds = []
        for cmd, args, kwargs in cmdify._commands.values():
            cmds.extend(parse_doc_for_commands(cmd.__doc__))

        for cmd, args, kwargs in cmdify._main_commands.values():
            cmds.extend(parse_doc_for_commands(cmd.__doc__))

        cwd = os.getcwd()
        os.chdir('commandify')
        for cmd in cmds:
            print(cmd)
            ret = subprocess.call(cmd, shell=True)
            assert ret == 0
        os.chdir(cwd)


commandify/version.py
VERSION = (0, 0, 4, 6, 'alpha')


def get_version(form='short'):
    if form == 'short':
        return '.'.join([str(v) for v in VERSION[:4]])
    elif form == 'long':
        return '.'.join([str(v) for v in VERSION][:4]) + '-' + VERSION[4]
    else:
        raise ValueError('unrecognised form specifier: {0}'.format(form))


def compare_to_current_version(version_str, form='short'):
    version = parse_version(version_str, form)
    return compare_versions(version, VERSION)


def parse_version(version_str, form='short'):
    project_status = VERSION[4]

    if form == 'long':
        version_numbers_str, project_status = version_str.split('-')
    elif form == 'short':
        version_numbers_str = version_str
    else:
        raise ValueError('Unrecognised form specifier: {0}'.format(form))
    try:
        version_numbers = [int(n) for n in version_numbers_str.split('.')]
    except ValueError:
        raise ValueError('All version numbers must be integers: "{0}"'
                         .format(version_str))

    if len(version_numbers) != 4:
        raise ValueError('Wrong length of version string: "{0}"'
                         .format(version_str))

    return tuple(version_numbers) + (project_status, )


def compare_versions(version1, version2):
    if version1 == version2:
        return 0
    return 1 if version1[:4] > version2[:4] else -1


__version__ = get_version()


if __name__ == '__main__':
    print(get_version())


commandify/commandify_examples.py
#!/usr/bin/env python
# Only required if you want to use argcomplete:
# PYTHON_ARGCOMPLETE_OK
'''Examples of a simple set of functions that use commandify
    usage::

        commandify_examples --help
        commandify_examples <command> --help
        commandify_examples --main-arg=22 <command>
        commandify_examples -m=22 <command>
        commandify_examples cmd_no_args

'''
from commandify import commandify, command, main_command


@main_command(main_arg={'flag': '-m'})
def main(main_arg=345):
    '''Example of how to use commandify
    Main command, called before any other command
    Can be used to set any global variables before any other commands executed.
    usage::

        commandify_examples --help
        commandify_examples --main-arg=22 cmd_no_args
        commandify_examples -m=22 cmd_no_args
        commandify_examples cmd_no_args

    '''
    print("Main command: {0}".format(main_arg))
    print(type(main_arg))

    return main_arg


@command
def cmd_no_args():
    '''Simplest command
    usage::

        commandify_examples cmd_no_args

    '''
    print('cmd_no_args called')


@command
def cmd1(name):
    '''Simple command with argument
    usage::

        commandify_examples cmd1 --name=a1
        commandify_examples cmd1 --name a3
        commandify_examples cmd1 --name 'a4'

    '''
    print('cmd1 running')
    print(name)


@command(username={'flag': '-n'})
def cmd2(username, userid):
    '''Command with 2 args and a short flag set for name
    usage::

        commandify_examples cmd2 --username=steve --userid=55
        commandify_examples cmd2 -n=sarah --userid=56

    '''
    print('{0}: {1}'.format(userid, username))


@command(some_arg={'default': 377})
def cmd3(some_arg, arg_with_default='arg_default'):
    '''Command with defaults set in two ways
    first through decorator and second through function arguments.
    usage::

        commandify_examples cmd3
        commandify_examples cmd3 --arg-with-default=default_overridden

    '''
    # Type of some_arg will be int:
    print(type(some_arg))
    print(some_arg, arg_with_default)


@command
def cmd4(some_arg=False):
    '''Command with False bool default
    usage::

        commandify_examples cmd4
        commandify_examples cmd4 --some-arg

    '''
    # Type of some_arg will be bool:
    print(type(some_arg))
    # If it is called specifying --some-arg, it will be True.
    print(some_arg)


@command
def cmd5(some_arg=True):
    '''Command with True bool default
    Command line argument gets turned into negative to handle this.
    usage::

        commandify_examples cmd5
        commandify_examples cmd5 --not-some-arg

    '''
    # Type of some_arg will be bool:
    print(type(some_arg))
    # If it is called without specifying --not-some-arg, it will be True.
    print(some_arg)


@command
def cmd6():
    # Command with no docstring.
    print('cmd6 called')


@command
def cmd7(args):
    '''All commands can easily access argparse args object
    usage::

        commandify_examples cmd7

    '''
    print('cmd7 called')
    print(args)


@command
def cmd8(args):
    '''All commands can access main command return value through args
    usage:
        commandify_examples cmd8
    '''
    print('cmd8 called')
    print(args.main_ret)
    return args.main_ret


@command
def cmd9(some_arg=23.):
    '''Command values are returnd via commandify(...)
    usage::

        commandify_examples cmd9

    '''
    return some_arg**2


if __name__ == '__main__':
    try:
        import argcomplete
        main_ret, command_ret =\
            commandify(suppress_warnings=['default_true'],
                       use_argcomplete=True,
                       exit=False)
        print('Main return: {0}'.format(main_ret))
        print('Command return: {0}'.format(command_ret))
    except ImportError:
        main_ret, command_ret = commandify(suppress_warnings=['default_true'],
                                           exit=False)
        print('Main return: {0}'.format(main_ret))
        print('Command return: {0}'.format(command_ret))
